runtime/lua/vim/lsp/protocol.lua

```diff
@@ -1250,14 +941,5 @@ protocol.Methods = {
   --- The `workspace/workspaceFolders` is sent from the server to the client to fetch the open workspace folders.
   workspace_workspaceFolders = 'workspace/workspaceFolders',
 }
-local function freeze(t)
-  return setmetatable({}, {
-    __index = t,
-    __newindex = function()
-      error('cannot modify immutable table')
-    end,
-  })
-end
-protocol.Methods = freeze(protocol.Methods)
```

Comment:
why not freeze?

Reply:
It makes no sense to freeze this, everything else in `vim.*` is free to mutate. Doesn't make any sense to me to single this one out. If someone wants to mutate this list for whatever reason (including adding a method), then I don't see why that shouldn't be allowed.

I believe in general that attempting to implement immutable objects in Lua is quite an anti pattern as it always comes with compromises.

Reply:
Different take: It makes no sense to mutate this.
Users can modify the result of `make_client_capabilities`, but the underlying constants should never change.

> Doesn't make any sense to me to single this one out

True, but why don't we go the other direction and freeze more components?

> If someone wants to mutate this list for whatever reason (including adding a method), then I don't see why that shouldn't be allowed.

Because it can lead to hard to track down bugs. We've had bug reports where some plugin was mutating something without the users being aware.

> I believe in general that attempting to implement immutable objects in Lua is quite an anti pattern as it always comes with compromises.

What makes it an anti-pattern?

Reply:
> True, but why don't we go the other direction and freeze more components?

To what real gain other than principal.

If we want immutability, then we need a language that properly accommodates it.

Reply:
> And it doesn't really make it immutable since the user can change the metatable anyway.

That's a fair point. As for the others...


> Also metatables are much slower than regular access.

I don't see any performance difference if only `__newindex` is set.

> For one it makes the table opaque, unable to inspect


`vim.inspect` works:

```
{
  x = 10,
  y = 20,
  <metatable> = {
    __newindex = <function 1>
  }
}
```

> and unable to iterate.

Not sure I follow, this shows only the keys of the original table?

```lua
local frozen_mt = {
  __newindex = function () error("Table is frozen") end
}
local p2 = setmetatable({x = 10, y = 20}, frozen_mt)
for key, value in pairs(p2) do
  print(key, value)
end
```

Output:

```
x 10
y 20
```


> Copying also becomes less straightforward.


`vim.deepcopy` copies the table, and assigns the metatable to the result.